Add termination logging to worker, refs #13504

mcantelon opened the pull request:
Added signal and shutdown handlers to the job worker so that termination can be logged.

mcantelon committed:
Add termination logging to worker, refs #13504

Added signal and shutdown handlers to the job worker so that
termination can be logged.

mcantelon commented:
I'm only intercepting INT, HUP, and TERM signals... not sure if I should intercept any others (there's about 30 of them).

jraddaoui commented on lib/task/jobs/jobWorkerTask.class.php:
```diff
@@ -119,6 +121,9 @@ function ($handle, $job, $e) {
             // could be disabled for this worker. See issue #4182.
             function () use (&$counter) {
+                // Needed for signal handler to work
+                pcntl_signal_dispatch();
```
Is it needed inside the loop? I'm not sure why it would be pending signals. Could it be solved calling `pcntl_async_signals(true);` only once. I'm not sure if they are related but I had to use the later to make signals work one time.

If not, could it be done with a worker callback?

jraddaoui commented on lib/task/jobs/jobWorkerTask.class.php:
```diff
@@ -127,4 +132,43 @@ function () use (&$counter) {
+        // to access the object's methods, etc.
+        $signalHandler = function ($signal) {
+            switch ($signal) {
+                case SIGINT:
```
I'd include SIGQUIT in here too.

jraddaoui commented on lib/task/jobs/jobWorkerTask.class.php:
```diff
@@ -127,4 +132,43 @@ function () use (&$counter) {
+                    break;
+
+                case SIGHUP:
+                    $this->log('Job worker restart requested.');
```
Is this logging okay ... it looks to be a termination signal caused by hang up ...

https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html

jraddaoui replied:
If that's right, we may want to call `exit()` in all cases (removing the breaks and adding a default case).

jraddaoui commented on lib/task/jobs/jobWorkerTask.class.php:
```diff
@@ -127,4 +132,43 @@ function () use (&$counter) {
+
+        pcntl_signal(SIGINT, $signalHandler);
+        pcntl_signal(SIGHUP, $signalHandler);
+        pcntl_signal(SIGTERM, $signalHandler);
```
Add `SIGQUIT`.

jraddaoui commented on lib/task/jobs/jobWorkerTask.class.php:
```diff
@@ -127,4 +132,43 @@ function () use (&$counter) {
+            $this->log('Job worker stopped.');
+        };
+
+        register_shutdown_function($shutdownFunction);
```
Nitpick: you could declare the function directly in the argument without using a variable.

jraddaoui reviewed (changes requested):
Nice @mcantelon, thanks!

I have added a few comments below. Also, this won't work in the Docker environment without installing the `pcntl` extension and changing how the signals are managed in the entrypoint script. I made those changes in the following commit for another work:

https://github.com/artefactual/atom/commit/9005f4551a2ef74366c2ff3f61a3c4e85824de53

You will only need to add the changes from `Dockerfile` and `docker/entrypoint.sh` to make it work, and here more info if you're interested:

https://www.ctl.io/developers/blog/post/gracefully-stopping-docker-containers/
https://medium.com/@gchudnov/trapping-signals-in-docker-containers-7a57fdda7d86
https://blog.forma-pro.com/dockerize-any-php-project-a-pure-bash-script-38dd37baf710
https://hynek.me/articles/docker-signals/
https://unix.stackexchange.com/questions/146756/forward-sigterm-to-child-in-bash
https://petermalmgren.com/signal-handling-docker/

At some point we may want to use `dumb-init`, but it's not required for this.

https://github.com/Yelp/dumb-init

mcantelon commented:
Hi @radda ... thanks for the tips! That first article "Gracefully Stopping Docker Containers" was super helpful!

I've updated the PR based on your feedback (and simplified the signal handler function a bit) and tested it in Docker... let me know what you think! Should we have different error codes for the different signal types?

jraddaoui commented:
Thanks @mcantelon, looks great!

If you don't mind, could you grab all the changes from the entrypoint?

https://github.com/artefactual/atom/commit/9005f4551a2ef74366c2ff3f61a3c4e85824de53#diff-4f5cabe26761257a4d685a6edc7a43e0fe0f78762f50eeb48530f2bd3b3ee7ca

It would be nice to have those clean-ups in here too, now that you're changing the worker command, it would be great to consolidate it for the fpm command.

PS. Sorry @Radda, I know it's not the first time, heheh. @mcantelon ... use @jraddaoui :P

mcantelon commented:
Hi @jraddaoui ... I put in all the `entrypoint.sh` changes and retested. Let me know if anything else should be tweaked!

jraddaoui commented:
Awesome, thanks @mcantelon!
